cpu/esp32/periph/adc_ctrl.c

```diff
@@ -0,0 +1,253 @@
+    RTCIO_ADC_ADC1,      /* GPIO34 VDET_1 */
+    RTCIO_ADC_ADC2,      /* GPIO35 VDET_2 */
+    RTCIO_SENSOR_SENSE1, /* GPIO36 SENSOR_VP */
+    RTCIO_SENSOR_SENSE2, /* GPIO37 SENSOR_CAPP */
+    RTCIO_SENSOR_SENSE3, /* GPIO38 SENSOR_CAPN */
+    RTCIO_SENSOR_SENSE4, /* GPIO39 SENSOR_VN */
+};
+
+/* flags to indicate whether the according controller is already initialized */
+static bool _adc1_ctrl_initialized = false;
+static bool _adc2_ctrl_initialized = false;
+
+void _adc1_ctrl_init(void)
+{
+    /* return if already intialized */
+    if (_adc1_ctrl_initialized) {
+        return;
+    }
+
+    /* always power on */
+    SENS.sar_meas_wait2.force_xpd_sar = SENS_FORCE_XPD_SAR_PU;
+
+    /* power off LN amp */
+    SENS.sar_meas_wait2.sar2_rstb_wait = 2;
+    SENS.sar_meas_ctrl.amp_rst_fb_fsm = 0;
+    SENS.sar_meas_ctrl.amp_short_ref_fsm = 0;
+    SENS.sar_meas_ctrl.amp_short_ref_gnd_fsm = 0;
+    SENS.sar_meas_wait1.sar_amp_wait1 = 1;
+    SENS.sar_meas_wait1.sar_amp_wait2 = 1;
+    SENS.sar_meas_wait2.sar_amp_wait3 = 1;
+    SENS.sar_meas_wait2.force_xpd_amp = SENS_FORCE_XPD_AMP_PD;
+
+    /* SAR ADC1 controller configuration */
+    SENS.sar_read_ctrl.sar1_dig_force = 0;      /* SAR ADC1 controlled by RTC */
+    SENS.sar_meas_start1.meas1_start_force = 1; /* SAR ADC1 started by SW */
+    SENS.sar_meas_start1.sar1_en_pad_force = 1; /* pad enable bitmap controlled by SW */
+    SENS.sar_touch_ctrl1.xpd_hall_force = 1;    /* XPD HALL is controlled by SW */
+    SENS.sar_touch_ctrl1.hall_phase_force = 1;  /* HALL PHASE is controlled by SW */
+    SENS.sar_read_ctrl.sar1_data_inv = 1;       /* invert data */
+    SENS.sar_atten1 = 0xffffffff;               /* set attenuation to 11 dB for all pads
+                                                   (input range 0 ... 3,3 V) */
+    /* power off built-in hall sensor */
+    RTCIO.hall_sens.xpd_hall = 0;
+
+    /* set default resolution */
+    SENS.sar_start_force.sar1_bit_width = ADC_RES_12BIT;
+    SENS.sar_read_ctrl.sar1_sample_bit = ADC_RES_12BIT;
+
+    _adc1_ctrl_initialized = true;
+}
+
+void _adc2_ctrl_init(void)
+{
+    /* return if already intialized */
+    if (_adc2_ctrl_initialized) {
+        return;
+    }
+
+    /* SAR ADC2 controller configuration */
+    SENS.sar_read_ctrl2.sar2_dig_force = 0;     /* SAR ADC2 controlled by RTC not DIG*/
+    SENS.sar_meas_start2.meas2_start_force = 1; /* SAR ADC2 started by SW */
+    SENS.sar_meas_start2.sar2_en_pad_force = 1; /* pad enable bitmap controlled by SW */
+    SENS.sar_read_ctrl2.sar2_data_inv = 1;      /* invert data */
+    SENS.sar_atten2 = 0xffffffff;               /* set attenuation to 11 dB for all pads
+                                                   (input range 0 ... 3,3 V) */
+    /* set default resolution */
+    SENS.sar_start_force.sar2_bit_width = ADC_RES_12BIT;
+    SENS.sar_read_ctrl2.sar2_sample_bit = ADC_RES_12BIT;
+
+    _adc2_ctrl_initialized = true;
+}
+
+extern const gpio_t _gpio_rtcio_map[];
```

Comment:
This is not needed right? It's defined before.

Reply:
of course